### DISCUSSION TITLE
ZigbeeModule with Ptvo to Sonoff Bridge back to ptvo module over rules

### DISCUSSION THREAD

**Discussion by mbmichi** [arendst/Tasmota#13804] [Q&A] [answered]
Hi, I would like to control the I/Os of a Zigbee Module with ptvo firmware over rules in a Sonoff ZBBridge. This is what I get, when a button is pressed:
`SENSOR = {"ZbReceived":{"0x1EED":{"Device":"0x1EED","MultiInValue":1,"Endpoint":2,"LinkQuality":60}}}`

Where Endpoint is the input number and MultiInValue is the way button was pressed (1..single, 2..double, etc.)
When I want to toggle an output pin I have to send i.e.:
`ZbSend { "device":"0x00124B00220942D2","Endpoint":3, "Send":{"Power":"TOGGLE"}}`
My problem is to find the correct syntax for a rule, which is sending that command automatically. I enabled it but it is never triggering.
Any idea about that?

**Comment by sfromis** [comment, marked as answer]
Well, what did you try?

**Reply by ascillato2** [reply]
> My problem is to find the correct syntax for a rule, which is sending that command automatically. I enabled it but it is never triggering.
> Any idea about that?

Please, share your rule to have an "idea" what can be wrong in your rule.

**Comment by mbmichi** [comment]
I could manage to find a trigger, which detects the button pressed with following rule:

```
Rule1 
	ON ZBReceived#?#Name DO var1 %value% ENDON 
	ON ZBReceived#?#Endpoint DO var2 %value% ENDON  
	ON ZBReceived#?#MultiInValue DO var3 %value% ENDON
	ON ZBReceived#?#MultiInValue DO IF ((%var1%=QM_CAR) AND (%var2%==2) AND (%var3%>0)) SENDCMD1 ENDIF ENDON
```

Is there an easier and shorter way to do that?

**Reply by sfromis** [reply]
You should really take a [look](https://tasmota.github.io/docs/Commands/#zigbee) at the various SetOption for Zigbee. SetOption83/ZbNameKey allows you to replace the use of that impractical "short address" by your ZbName setting, and SetOption100 /ZbNoPrefix allows you to remove the ZbReceived object level in the JSON. There are also other useful possibilities depending on your preference.

Using the two options suggested above, I can have a rule like 
`on Ikea-Opal-1#Dimmer>200 do ZbSend {"Device":"Ikea-Opal-2","Send":{"Dimmer":%value%}} endon`

**Comment by mbmichi** [comment]
and when I replace SENDCMD1 with the zbsend command, it is working like it should. However the rule needs 322Bytes and I have just 190Bytes free. So I can add just 1 rule like this.

```
Rule1 
	ON ZBReceived#?#Name DO var1 %value% ENDON 
	ON ZBReceived#?#Endpoint DO var2 %value% ENDON  
	ON ZBReceived#?#MultiInValue DO var3 %value% ENDON
	ON ZBReceived#?#MultiInValue DO IF ((%var1%=QM_CAR) AND (%var2%==2) AND (%var3%>0)) ZbSend { "device":"0x00124B00220942D2","Endpoint":2, "Send":{"Power":"TOGGLE"}} ENDIF ENDON
```

**Reply by sfromis** [reply]
When looking at bytes used in rule sets, do notice messages like this:
`RUL: Stored uncompressed, would compress from 79 to 44 (-44%)`
When rule sets grow, this compression will be used to reduce the size, for longer ones you can expect compression well above 50%. Your 3 rule sets of each 512 bytes would allow over 3KB of rule text.

In your rule set above, you can greatly simplify using the technique from my other response, but still, you could replace the two last rules by one, by using %value% directly in the last rule, instead of %var3%. 

Of course, my other suggestion would allow a trigger like
`ON QM-CAR#MultiInValue DO ...`

More generally, you can use the Event command and triggers to "parcel out" actions from one rule to another. This may be convenient for many reasons.

**Comment by mbmichi** [comment]
Thanks for the replies. I will check tomorrow. However size is very limited. I would need more then 2 rules like this. Is there an alternative to sonoff zb-bridge with more space for rules?

**Reply by sfromis** [reply]
Not to get more than about 3KB total space of the 3 rule sets. You can get quite a few rules (around 8 of your example) into those, especially if you use the improved settings allowing much simpler rules.

However there is a [very similar bridge based on ESP32](https://templates.blakadder.com/ewelink_ZB-GW03.html), where you have the Berry language available, with greatly expanded capacity and features compared to rules.

**Comment by mbmichi** [comment]
So when I get this response, 
```
`{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":322,"Free":189,"Rules":"ON ZBReceived#?#Name DO var1 %value% ENDON  \tON ZBReceived#?#Endpoint DO var2 %value% ENDON   \tON ZBReceived#?#MultiInValue DO var3 %value% ENDON \tON ZBReceived#?#MultiInValue DO IF ((%var1%=QM_CAR) AND (%var2%==2) AND (%var3%>0)) ZbSend { \"device\":\"0x00124B00220942D2\",\"Endpoint\":2, \"Send\":{\"Power\":\"TOGGLE\"}} ENDIF ENDON"}}`
```
free 189Bytes does not mean the free space for other rules. Up to 8 rules would be enough. I will have a look to the other bridge aswell. Many thanks for your help.

**Reply by sfromis** [reply]
189 bytes means free space in one of the 3 rule sets. It does not mean how much you can add to that one rule set, due to compression kicking in, which for more than short rules normally at least doubles bytes you can have.
